### PR TITLE
frontend: data-pipelines publish artifact CI/CD

### PR DESCRIPTION
**Pull request by ivakoleva**
frontend: data-pipelines publish artifact CI/CD

We need to introduce a script for publishing NPM packages, with the
account we have provisioned for https://registry.npmjs.org/.

Added a script that supports publishing NPM artifacts of frontend 
subprojects in [shared-components data-pipelines].
Added a version.txt file for `data-pipelines/gui`, reused the mechanism
introduced to shared for bumping minor version with $CI_PIPELINE_ID.
Reused the NPM registry auth, and cleaned up the obsolete script
publish_shared.sh in favour of publish_package_npm.sh since 
more generic.

Testing done: run the publish script to test using --dry-run; ci/cd
See successful publishing dry-run https://gitlab.com/vmware-analytics/versatile-data-kit/-/pipelines/797108389

### CODE REVIEW

File: projects/frontend/cicd/.gitlab-ci.yml [hunk 1]
```diff
@@ -1,6 +1,59 @@
...
+  before_script:
+    - cd projects/frontend/
+  script:
+    - ./cicd/publish_package_npm.sh data-pipelines
+  retry: !reference [.retry, retry_options]
+  rules:
+    - if: '$CI_PIPELINE_SOURCE == "schedule"'
+      when: never
+    # TODO: cleanup external_pull_request_event before merge; for dev verification purposes only
+    - if: '$CI_COMMIT_BRANCH == "main" || $CI_PIPELINE_SOURCE == "external_pull_request_event"'
```
Comment (ivakoleva) [inline projects/frontend/cicd/.gitlab-ci.yml:48]: will cleanup this todo before merge; 
external_pull_request_event used to replay the entire workflow in advance for review

File: projects/frontend/cicd/publish_package_npm.sh [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    exit 3
+fi
+project_type=$1
+cd "../$project_type/gui"
+
+# TODO: generated a token, and added NPM_TOKEN to gitlab ci/cd variables, need to login
+# Login
+#npm-cli-login -u ${NPM_USERNAME} \
+#              -p ${NPM_PASSWORD} \
+#              -e "versatiledatakit@groups.vmware.com"
```
Comment (ivakoleva) [inline projects/frontend/cicd/publish_package_npm.sh:42]: this is WIP

File: projects/frontend/data-pipelines/gui/projects/data-pipelines/tsconfig.lib.prod.json [hunk 3]
```diff
@@ -5,8 +5,7 @@
     "declarationMap": false
   },
   "angularCompilerOptions": {
-    "compilationMode": "full",
-    "enableIvy": true,
+    "compilationMode": "partial",
```
Comment (sabadzhiev) [inline projects/frontend/data-pipelines/gui/projects/data-pipelines/tsconfig.lib.prod.json:8]: Just a note:  
 - The packaging using Ivy and this compilation mode were set in this manner to resolve loading issues while consuming the library in the end application.
 Just to have in mind, if the consuming project have troubles, to look into this.

Comment (ivakoleva) [reply projects/frontend/data-pipelines/gui/projects/data-pipelines/tsconfig.lib.prod.json]: The publishing in dry-run using full Ivy mode was successful https://gitlab.com/vmware-analytics/versatile-data-kit/-/pipelines/797108389, closing the conversation.

Last time, the npm registry discarded the package compiled in --dry-run for full Ivy mode. Rolled back the full Ivy mode, keeping less diverging as passible for easier adoption back.

File: projects/frontend/cicd/publish_package_npm.sh [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+fi
+project_type=$1
+cd "../$project_type/gui"
+
+if [ $# -eq 1 ]
+  then
+    echo "ERROR: No argument for unique minor package version provided."
+    exit 3
+fi
+minor_version_unique=$2
```
Comment (DeltaMichael) [inline projects/frontend/cicd/publish_package_npm.sh:48]: This should be called `patch_version`.

File: projects/frontend/cicd/publish_package_npm.sh [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+###
+# This script is purposed for publishing to https://registry.npmjs.org/ (default public repo).
+# It is meant to publish the release NPM artifact of a subproject in `projects/frontend/`:
+# * shared-components
+# * data-pipelines
+#
+# Usage:
+# ./publish_package_npm.sh <component> <minor_version_unique> <token> [npm_registry_optional]
```
Comment (DeltaMichael) [inline projects/frontend/cicd/publish_package_npm.sh:13]: `minor_version_unique` should be `patch_version_unique`

**Review by DeltaMichael — APPROVED** [review]

**Pull request merged**